### ISSUE TITLE
event.DeleteEvent struct's "DeleteStateUnknown" field is never written

### ISSUE THREAD

**Issue by seh** [kubernetes-sigs/controller-runtime#654]
[The `event.DeleteEvent` type](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/event#DeleteEvent) has a field called "DeleteStateUnknown," documented as follows:
> DeleteStateUnknown is true if the Delete event was missed but we identified the object as having been deleted.

At present, this field is never written to, and nothing in the _controller-runtime_ library reads it either. This field entered the source back in #2, on 7 June 2018, copying the code from [the _kubernetes-sigs/kubebuilder_ project](https://github.com/kubernetes-sigs/kubebuilder).

Per [discussion in the "kubebuilder" channel](https://kubernetes.slack.com/archives/CAR30FCJZ/p1571518415160500) in the "Kubernetes" Slack team, @pwittrock suggested that perhaps we should use this field, writing to it when appropriate, rather than removing it from the library, which would require a major version change to avoid breaking current `Predicate`s.

Not being sure of the original intention for this field, it looks like [the `source/internal.EventHandler.OnDelete` method](https://sourcegraph.com/github.com/kubernetes-sigs/controller-runtime@master/-/blob/pkg/source/internal/eventsource.go#L148) might be a fitting place, to indicate that we found a `cache.DeleteFinalStateUnknown` object in the cache, indicating that we missed a deleting watch event and might therefore have only a stale object available.

For `Reconciler`s, if I understand the contract correctly, the distinction seems immaterial, as by the time they would go to inspect the object, it would be gone anyway; only its name and possibly its containing namespace would be presented to the `Reconciler`. However, other event handlers—and [`Predicates` that precede them](https://sourcegraph.com/github.com/kubernetes-sigs/controller-runtime@master/-/blob/pkg/source/internal/eventsource.go#L169-L173)—would get a chance to see this tombstone object.

Should we populate this field? If so, where and when? Does its documentation adequately explain its meaning?

**Comment by DirectXMan12** [comment]
can we get some context from k/k?  I don't remember what our intention was with that field, but the tombstone logic sits in a bit of a minefield of weird artifacts ported from k/k.

**Comment by vincepri** [comment]
@seh Were you able to get more information from k/k?

**Comment by seh** [comment]
No, sorry, I didn't find anything telling when I last looked into it.

**Comment by Mitsuwa** [comment]
Does this field actually do anything?  I have also noticed that it doesn't seem to populate
